scripts/37-functional-rust.md

```diff
@@ -0,0 +1,1189 @@
+
+It's got the practicality of a chainsaw, but what I want is a poem that cuts just as sharp.
+
+---
+
+![[rust-logo.png]]
+
+notes:
+
+in 2020, after a few false starts I found rust.
+
+And I have great news, Rust is more functional than I realised until recently.
+
+The functional ML roots of the language, Graydon's first Rust compiler was written in OCaml, shine through, influencing it right from the start.
+
+It's not "C++ but better".
+
+It's Haskell standing on Lisp's shoulders, hiding in C's coat to sneak into _PRDCTN_. (The fancy nightclub where all the popular language's hang out)
+
+Let's look at how Rust can manage side effects
+
+---
+
+# Set Up
+
+```toml
+[dependencies]
+color-eyre = "0.6.2"
+rstest = "0.18.2"
+```
+
+(in `cargo.toml`)
+
+notes:
+
+Here is the classic set-up I include in all of my Rust projects, pretty and simple error handling with `color-eyre`, and the `rstest` testing framework, the most important feature for me is its ability to use fixtures.
+
+---
+
+## Killer Crates
+
+```toml
+rayon = "1.8"
+const_panic = "0.2"
+```
+
+notes:
+
+Later on, I'm going to use these two cool crates to demo the utility of Rust's functional techniques today.
+- `Rayon` is the simplest parallelism library you've ever seen, built on Rust's controlled side-effect systems, and
+- `const_panic` allows string formatting for asserts in a const context.
+
+---
+
+## Superpower Lints
+
+```toml
+[lints.rust]
+unsafe_code = "forbid"
+
+[lints.clippy]
+enum_glob_use = "deny"
+pedantic = "deny"
+nursery = "deny"
+unwrap_used = "deny"
+```
+
+```sh
+$ cargo clippy
+$ bacon clippy # or even better!
+
+[clippy negs your code here]
+```
+
+notes:
+
+Enable all these lints. Allow individual lines if you need to.
+
+If you're using rust without using `bacon` and `clippy`, you're doing yourself a great disservice.
+
+VSCode or your editor will likely use clippy behind the scenes, but you must also run clippy in a terminal to answer the most important question:
+
+"What was the first error", as that's the one you have to fix.
+
+---
+
+```toml
+[profile.release]
+opt-level = 'z'   # Optimize for size.
+lto = true        # Enable Link Time Optimisation
+codegen-units = 1 # Reduced to increase optimisations.
+panic = 'abort'   # Abort on panic
+strip = "symbols" # Strip symbols from binary
+```
+
+More: https://github.com/johnthagen/min-sized-rust
+
+notes:
+
+- [ ] retake this slide
+
+And, as a bonus, here's a reasonable starting point for optimising the size of your release builds, cutting a helloworld down from 3MB to 300K.
+
+You can get it down to 8K if you really want to, but I recommend stopping here and starting building.
+
+---
+
+```rust
+// #[cfg(test)] omitted because I want linting
+mod tests {
+
+#[allow(clippy::wildcard_imports)]
+use rstest::*;
+
+#[fixture]
+fn setup() {
+	color_eyre::install().expect("color_eyre installed");
```

Comment:
```suggestion
	color_eyre::install().unwrap_or_else(|report| std::process::exit(<report as std::process::Termination>.report()));
```

Probably the better idea is to just put "could not install eyre" in the message, but i wanted to know how to do this:
it uses the Termination trait to process the report in a status code and exits the process with it. this is what Rust does in the background when you return a Result<(), Report> from `main` afaict

Reply:
oh nice! I might just allow an exception to `unwrap_used` here and call it a day with an `.unwrap()` (for brevity)

Reply:
Also, @laundmo, are you in the "errors go in expect() messages" camp, not "my expectations go in expect() messages" camp?!

I'm astonished :wink: https://stackoverflow.com/questions/66362625/why-is-rusts-expect-called-expect

Reply:
i am in the "the error i expect to trigger this" camp